### DISCUSSION TITLE
Customizing the autogenerated __init__ function of SQLAlchemy ORM models with Mapped columns while preserving type hinting

### DISCUSSION THREAD

**Discussion by jerem1e** [sqlalchemy/sqlalchemy#9768] [Usage Questions]
I am working with SQLAlchemy ORM models that use Mapped columns, and I would like to customize the autogenerated __init__ function to have a specific signature while preserving type hinting. Here's an example of the model:

```python
class User(Base):
    __tablename__ = "users"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column()
    age: Mapped[int] = mapped_column()
```
The autogenerated `__init__` function currently has the generic signature `__init__(*args, **kwargs)`, which makes it challenging to enforce type safety. However, I would like the ` __init__` function to have the following signature:

```python
def __init__(self, id: int, name: str, age: int) -> None:
    self.id = id
    self.name = name
    self.age = age
```
Is there a way to customize the autogenerated __init__ function of SQLAlchemy ORM models with Mapped columns to match the desired signature while preserving type hinting? I would like to take advantage of type checking tools, such as mypy, to ensure type correctness during development.

**Comment by CaselIT** [comment]
Hi,

You can use the MappedAsDataclass to map the classes as dataclasses

**Reply by zzzeek** [reply]
link: https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#declarative-dataclass-mapping    

this is what we offer for type-aware `__init__` methods.   other approaches you'd want to roll your own using an `__init_subclass__()` hook.
